Link error

## Expected Behavior

Build without link error.

## Actual Behavior

When using HPX CMake targets it seems pthread is not in the list of libs to link against.


## Steps to Reproduce the Problem

  1. Configure HPX without any arguments.
  1. Compile a simple Hello World program
  1. Link against hpx, wrap_main, iostreams_component

## Specifications

  - HPX Version: 1.7.0-rc2
  - Platform (compiler, OS): gcc / linux (kubuntu 20.04)

It seems relevant that the libasio-dev package is used to provide asio. When configuring HPX with `HPX_WITH_FETCH_ASIO=ON` all is well.

When passing `pthread` to CMake's `target_link_libraries` (in addition to HPX' targets), the link error is gone.

Adding`-DPTHREAD_PREFER_PTHREAD_FLAG=TRUE` resulted in the same link error.

Adding`-DPTHREAD_PREFER_PTHREAD_FLAG=TRUE` and commenting out the two lines in `hpx/CMakeLists.txt` resulted in a CMake configure error:

```
-- Configuring done
CMake Error at /tmp/hello_hpx-fetch_content/Debug-default/build/_deps/hpx-src/cmake/HPX_GeneratePackageUtils.cmake:368 (file):
  Error evaluating generator expression:

    $<COMPILE_LANG_AND_ID:CUDA,NVIDIA>

  $<COMPILE_LANG_AND_ID:lang,id> may only be used with binary targets to
  specify include directories, compile definitions, and compile options.  It
  may not be used with the add_custom_command, add_custom_target, or
  file(GENERATE) commands.
Call Stack (most recent call first):
  /tmp/hello_hpx-fetch_content/Debug-default/build/_deps/hpx-src/cmake/HPX_GeneratePackage.cmake:145 (hpx_generate_pkgconfig_from_target)
  /tmp/hello_hpx-fetch_content/Debug-default/build/_deps/hpx-src/CMakeLists.txt:2533 (include)
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1730,8 +1730,6 @@ if(NOT WIN32)
   # System libraries
   # ############################################################################
   if(NOT MSVC)
-    hpx_add_compile_flag_if_available(-pthread)
-    hpx_add_link_flag_if_available(-pthread TARGETS EXE SHARED)
     if(HPX_CXX11_STD_ATOMIC_LIBRARIES)
       target_link_libraries(
         hpx_base_libraries INTERFACE ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
@@ -2146,6 +2144,8 @@ endif()
 # Find Our dependencies: These are all dependencies needed to build the core
 # library. Dependencies that are only needed by plugins, examples or tests
 # should be found separately in the appropriate subdirectory.
+
+include(HPX_SetupThreads)
 
 # Setup our required Boost libraries.
 include(HPX_SetupBoost)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -323,7 +323,10 @@ if(HPX_WITH_PRECOMPILED_HEADERS)
   set(HPX_WITH_PRECOMPILED_HEADERS_INTERNAL ON)
 
   add_library(hpx_precompiled_headers OBJECT libs/src/dummy.cpp)
-  target_link_libraries(hpx_precompiled_headers PRIVATE hpx_private_flags)
+  target_link_libraries(
+    hpx_precompiled_headers PRIVATE hpx_public_flags hpx_private_flags
+                                    hpx_base_libraries
+  )
   target_precompile_headers(
     hpx_precompiled_headers
     PRIVATE

--- a/cmake/FindAsio.cmake
+++ b/cmake/FindAsio.cmake
@@ -7,7 +7,6 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-hpx_warn("finding asio")
 if(NOT TARGET Asio::asio)
   find_path(
     ASIO_INCLUDE_DIR asio.hpp

--- a/cmake/HPX_GeneratePackage.cmake
+++ b/cmake/HPX_GeneratePackage.cmake
@@ -141,19 +141,21 @@ if(HPX_WITH_PKGCONFIG)
     hpx_pkgconfig_component INTERFACE "-std=c++${HPX_CXX_STANDARD}"
   )
 
+  set(exclude_targets hpx_interface Threads::Threads)
+
   # Generate the pkconfig files for HPX_APPLICATION (both for build and install)
   hpx_generate_pkgconfig_from_target(
-    hpx_pkgconfig_application hpx_application TRUE EXCLUDE hpx_interface
+    hpx_pkgconfig_application hpx_application TRUE EXCLUDE ${exclude_targets}
   )
   hpx_generate_pkgconfig_from_target(
-    hpx_pkgconfig_application hpx_application FALSE EXCLUDE hpx_interface
+    hpx_pkgconfig_application hpx_application FALSE EXCLUDE ${exclude_targets}
   )
   # Generate the pkconfig files for HPX_COMPONENT (both for build and install)
   hpx_generate_pkgconfig_from_target(
-    hpx_pkgconfig_component hpx_component TRUE EXCLUDE hpx_interface
+    hpx_pkgconfig_component hpx_component TRUE EXCLUDE ${exclude_targets}
   )
   hpx_generate_pkgconfig_from_target(
-    hpx_pkgconfig_component hpx_component FALSE EXCLUDE hpx_interface
+    hpx_pkgconfig_component hpx_component FALSE EXCLUDE ${exclude_targets}
   )
 
   string(TOLOWER ${CMAKE_BUILD_TYPE} build_type)

--- a/cmake/HPX_SetupBoost.cmake
+++ b/cmake/HPX_SetupBoost.cmake
@@ -84,9 +84,6 @@ if(NOT TARGET hpx_dependencies_boost)
 
   include(HPX_AddDefinitions)
 
-  find_package(Threads QUIET REQUIRED)
-  target_link_libraries(hpx_dependencies_boost INTERFACE Threads::Threads)
-
   # Boost preprocessor definitions
   if(NOT Boost_USE_STATIC_LIBS)
     hpx_add_config_cond_define(BOOST_ALL_DYN_LINK)

--- a/cmake/HPX_SetupThreads.cmake
+++ b/cmake/HPX_SetupThreads.cmake
@@ -1,0 +1,13 @@
+# Copyright (c) 2018 Christopher Hinz
+# Copyright (c) 2014 Thomas Heller
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+find_package(Threads REQUIRED)
+
+if(NOT HPX_FIND_PACKAGE)
+  target_link_libraries(hpx_base_libraries INTERFACE Threads::Threads)
+endif()

--- a/cmake/templates/HPXConfig.cmake.in
+++ b/cmake/templates/HPXConfig.cmake.in
@@ -81,6 +81,8 @@ if(NOT MIMALLOC_ROOT AND NOT "$ENV{MIMALLOC_ROOT}")
 endif()
 include(HPX_SetupAllocator)
 
+include(HPX_SetupThreads)
+
 # Boost Separate boost targets to be unarily linked to some modules
 set(HPX_BOOST_ROOT "@BOOST_ROOT@")
 # By default BOOST_ROOT is set to HPX_BOOST_ROOT (not necessary for PAPI or

